includes/class-llms-awards-query.php

```diff
@@ -0,0 +1,340 @@
+		if ( 'any' === $type ) {
+			$types = $this->allowed_post_types;
+		} elseif ( 'achievement' === $type ) {
+			$types = array( 'llms_my_achievement' );
+		} elseif ( 'certificate' === $type ) {
+			$types = array( 'llms_my_certificate' );
+		}
+
+		return $types;
+
+	}
+
+	/**
+	 * Prepares the `meta_query` ultimately passed to the WP_Query.
+	 *
+	 * @since [version]
+	 *
+	 * @return array An array of meta query arrays.
+	 */
+	private function prepare_meta_query() {
+
+		// If a query for manual awards we skip all other relationships and return early.
+		if ( $this->get( 'manual_only' ) ) {
+
+			return array(
+				'relation' => 'OR',
+				// Meta doesn't exist.
+				array(
+					'key'     => '_llms_engagement',
+					'compare' => 'NOT EXISTS',
+				),
+				// Or it's "empty".
+				array(
+					'key'     => '_llms_engagement',
+					'value'   => array( '', '0', 0 ),
+					'compare' => 'IN',
+				),
+			);
+
+		}
+
+		return $this->prepare_meta_query_for_relationships();
+
+	}
+
+	/**
+	 * Retrieve meta query parts for related posts, engagements, and templates.
+	 *
+	 * @since [version]
+	 *
+	 * @return array An array of meta query arrays.
+	 */
+	private function prepare_meta_query_for_relationships() {
+
+		$meta_query = array();
+
+		$relations = array(
+			'related_posts' => '_llms_related',
+			'engagements'   => '_llms_engagement',
+		);
+		foreach ( $relations as $arg => $meta_key ) {
+
+			// Include.
+			if ( ! empty( $this->get( $arg ) ) ) {
+				$meta_query[] = array(
+					'key'     => $meta_key,
+					'value'   => $this->sanitize_id_array( $this->get( $arg ) ),
+					'compare' => 'IN',
+				);
+			}
+
+			// Exclude.
+			$exclude_arg = $arg . '__exclude';
+			if ( ! empty( $this->get( $exclude_arg ) ) ) {
+
+				$meta_query[] = array(
+					'relation' => 'OR',
+					array(
+						'key'     => $meta_key,
+						'value'   => $this->sanitize_id_array( $this->get( $exclude_arg ) ),
+						'compare' => 'NOT IN',
+					),
+					// Ensure posts that don't have the metadata set will be returned.
+					array(
+						'key'     => $meta_key,
+						'compare' => 'NOT EXISTS',
+					),
+				);
+			}
+		}
+
+		return $meta_query;
+
+	}
+
+	/**
+	 * Prepare the WP_Query arguments for the awards query.
+	 *
+	 * @since [version]
+	 *
+	 * @return array Array of arguments suitable to pass to a WP_Query.
+	 */
+	protected function prepare_query() {
+
+		// Remove any extra arguments not found in the map.
+		$args = array_intersect_key(
+			parent::prepare_query(),
+			array_flip( $this->get_arg_map() )
+		);
+
+		// Add post type(s).
+		$args['post_type'] = $this->post_types();
+
+		// Add meta query.
+		$args['meta_query'] = $this->prepare_meta_query();
+
+		foreach ( $args as $arg => $val ) {
+			if ( is_array( $val ) && empty( $val ) ) {
+				unset( $args[ $arg ] );
+			}
+		}
+
+		// Remove empty arrays.
+		return array_filter(
+			$args,
+			function( $val ) {
+				return ! is_array( $val ) || ! empty( $val );
+			}
+		);
```

Comment:
these two ( the `for` and the `array_filter` ) are equivalent, right?

Reply:
yea i forgot to delete the `foreach` after i rewrote it so you'd be impressed with my golfing skills...